overlord/snapstate/handlers_components.go

```diff
@@ -0,0 +1,229 @@
+	}
+	if err := ts.Get("component-setup", &compSetup); err != nil {
+		return nil, nil, err
+	}
+	return &compSetup, snapsup, nil
+}
+
+func (m *SnapManager) doPrepareComponent(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	compSetup, _, err := TaskComponentSetup(t)
+	if err != nil {
+		return err
+	}
+
+	if compSetup.Revision().Unset() {
+		// This is a local installation, revision is -1 (there
+		// is no history of local revisions for components).
+		compSetup.CompSideInfo.Revision = snap.R(-1)
+	}
+
+	t.Set("component-setup", compSetup)
+	return nil
+}
+
+func (m *SnapManager) doMountComponent(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	perfTimings := state.TimingsForTask(t)
+	compSetup, snapsup, err := TaskComponentSetup(t)
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	st.Lock()
+	deviceCtx, err := DeviceCtx(t.State(), t, nil)
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	// TODO we might want a checkComponents doing checks for some
+	// component types (see checkSnap and checkSnapCallbacks slice)
+
+	csi := compSetup.CompSideInfo
+	cpi := snap.MinimalComponentContainerPlaceInfo(csi.Component.ComponentName,
+		csi.Revision, snapsup.InstanceName(), snapsup.Revision())
+
+	cleanup := func() {
+		st.Lock()
+		defer st.Unlock()
+
+		// RemoveComponentDir is idempotent so it's ok to always
+		// call it in the cleanup path.
+		if err := m.backend.RemoveComponentDir(cpi); err != nil {
+			t.Errorf("cannot cleanup partial setup component %q: %v",
+				compSetup.CompSideInfo, err)
+		}
+	}
+
+	pm := NewTaskProgressAdapterUnlocked(t)
+	var installRecord *backend.InstallRecord
+	timings.Run(perfTimings, "setup-component",
+		fmt.Sprintf("setup component %q", compSetup.CompSideInfo.Component),
+		func(timings.Measurer) {
+			installRecord, err = m.backend.SetupComponent(
+				compSetup.CompPath,
+				cpi,
+				deviceCtx,
+				pm)
+		})
+	if err != nil {
+		cleanup()
+		return err
+	}
+
+	// double check that the component is mounted
+	var readInfoErr error
+	for i := 0; i < 10; i++ {
+		compMntDir := cpi.MountDir()
+		_, readInfoErr = readComponentInfo(compMntDir)
+		if readInfoErr == nil {
+			logger.Debugf("component %q (%v) available at %q",
+				compSetup.CompSideInfo.Component,
+				compSetup.Revision(), compMntDir)
+			break
+		}
+		// snap not found, seems is not mounted yet
+		time.Sleep(mountPollInterval)
+	}
+	if readInfoErr != nil {
+		timings.Run(perfTimings, "undo-setup-component",
+			fmt.Sprintf("Undo setup of component %q",
+				compSetup.CompSideInfo.Component),
+			func(timings.Measurer) {
+				err = m.backend.UndoSetupComponent(cpi,
+					installRecord, deviceCtx, pm)
+			})
+		if err != nil {
+			st.Lock()
+			t.Errorf("cannot undo partial setup of component %q: %v",
+				compSetup.CompSideInfo.Component, err)
+			st.Unlock()
+		}
+
+		cleanup()
+		return fmt.Errorf("expected component %q rev %v to be mounted but is not: %w",
```

Comment:
how does the corresponding error for snaps looks like?  "rev %v" is a bit odd, I don't think we use it frequently or at all?